File: Linked Lists/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  dummy = new ListNode(0, head);
  let behind = dummy,
    ahead = dummy;

  for (let i = 0; i < n + 1; i++) ahead = ahead.next;

  while (ahead) {
    ahead = ahead.next;
    behind = behind.next;
  }

  behind.next = behind.next.next;

  return dummy.next;
};

// Time: O(n)
// Space: O(1)

File: Linked Lists/19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        behind = ahead = dummy

        for _ in range(n + 1):
            ahead = ahead.next

        while ahead:
            ahead = ahead.next
            behind = behind.next

        behind.next = behind.next.next

        return dummy.next

# Time: O(n)
# Space: O(1)
